BINLOG_ADAPT_CONCEPT does not support namespaces

```cpp
namespace  ns {

template<typename T>
concept Concept2 = requires(T t)
{
    {t.A} -> std::convertible_to<std::string_view>;
    {t.B} -> std::convertible_to<std::string_view>;
};

} // namespace  ns

BINLOG_ADAPT_CONCEPT(ns::Concept2, A, B)
```

```
../../source/utils/test/test_serialize.cpp:33:22: error: template argument for template type parameter must be a type; did you forget 'typename'?
BINLOG_ADAPT_CONCEPT(ns::Concept2, A, B)
                     ^
                     typename
_deps/binlog-src/include/binlog/adapt_struct.hpp:153:79: note: expanded from macro 'BINLOG_ADAPT_CONCEPT'
  MSERIALIZE_EXPAND(MSERIALIZE_MAKE_TEMPLATE_SERIALIZABLE((Concept Concept), (Concept) __VA_OPT__(,) __VA_ARGS__)) \
                                                                              ^
_deps/binlog-src/include/mserialize/make_template_serializable.hpp:63:59: note: expanded from macro 'MSERIALIZE_MAKE_TEMPLATE_SERIALIZABLE'
    :StructSerializer<MSERIALIZE_UNTUPLE(MSERIALIZE_FIRST(__VA_ARGS__)) \
                                                          ^
_deps/binlog-src/include/mserialize/detail/preprocessor.hpp:8:68: note: expanded from macro 'MSERIALIZE_FIRST'
#define MSERIALIZE_FIRST(...) MSERIALIZE_EXPAND(MSERIALIZE_FIRST_I(__VA_ARGS__, 0))
                                                                   ^
note: (skipping 2 expansions in backtrace; use -fmacro-backtrace-limit=0 to see all)
_deps/binlog-src/include/mserialize/detail/preprocessor.hpp:20:52: note: expanded from macro 'MSERIALIZE_UNTUPLE'
#define MSERIALIZE_UNTUPLE(a) MSERIALIZE_UNTUPLE_I a
                                                   ^
_deps/binlog-src/include/mserialize/detail/preprocessor.hpp:21:35: note: expanded from macro 'MSERIALIZE_UNTUPLE_I'
#define MSERIALIZE_UNTUPLE_I(...) __VA_ARGS__
                                  ^
_deps/binlog-src/include/mserialize/detail/preprocessor.hpp:5:30: note: expanded from macro 'MSERIALIZE_EXPAND'
#define MSERIALIZE_EXPAND(x) x
                             ^
_deps/binlog-src/include/mserialize/StructSerializer.hpp:88:20: note: template parameter is declared here
template <typename T, typename... Members>
```

--- a/.clang-tidy
+++ b/.clang-tidy
@@ -10,38 +10,43 @@
 Checks: "\
   *,
   -altera-*,\
-  -fuchsia-*,\
+  -bugprone-easily-swappable-parameters,\
   -bugprone-sizeof-expression,\
+  -cert-err58-cpp,\
+  -cert-err60-cpp,\
   -cppcoreguidelines-avoid-c-arrays,\
   -cppcoreguidelines-avoid-magic-numbers,\
   -cppcoreguidelines-avoid-non-const-global-variables,\
+  -cppcoreguidelines-init-variables,\
   -cppcoreguidelines-macro-usage,\
-  -cppcoreguidelines-pro-type-vararg,\
-  -cppcoreguidelines-pro-type-reinterpret-cast,\
   -cppcoreguidelines-pro-bounds-array-to-pointer-decay,\
   -cppcoreguidelines-pro-bounds-constant-array-index,\
   -cppcoreguidelines-pro-bounds-pointer-arithmetic,\
-  -cppcoreguidelines-init-variables,\
+  -cppcoreguidelines-pro-type-reinterpret-cast,\
+  -cppcoreguidelines-pro-type-vararg,\
+  -fuchsia-*,\
+  -google-readability-casting,\
   -hicpp-*,\
-  -cert-err58-cpp,\
-  -cert-err60-cpp,\
-  -readability-convert-member-functions-to-static,\
-  -readability-function-cognitive-complexity,\
-  -readability-magic-numbers,\
-  -readability-named-parameter,\
-  -readability-uppercase-literal-suffix,\
-  -misc-non-private-member-variables-in-classes,\
+  -llvm-header-guard,\
   -misc-no-recursion,\
+  -misc-non-private-member-variables-in-classes,\
   -modernize-avoid-c-arrays,\
   -modernize-concat-nested-namespaces,\
   -modernize-return-braced-init-list,\
   -modernize-unary-static-assert,\
   -modernize-use-auto,\
+  -modernize-use-default-member-init,\
   -modernize-use-nodiscard,\
   -modernize-use-trailing-return-type,\
-  -llvm-header-guard,\
+  -readability-convert-member-functions-to-static,\
+  -readability-function-cognitive-complexity,\
+  -readability-identifier-length,\
+  -readability-magic-numbers,\
+  -readability-named-parameter,\
+  -readability-uppercase-literal-suffix,\
   -llvmlibc-*,\
   -google-runtime-references"
+
 
 HeaderFilterRegex: '.*binlog/.*|.*mserialize/.*'
 

--- a/.github/workflows/on_pull_request.yml
+++ b/.github/workflows/on_pull_request.yml
@@ -106,8 +106,6 @@ jobs:
       - uses: actions/checkout@v2
       - name: Install Dependencies
         run: |
-          # remove broken packages installed by default
-          sudo apt remove -y clang-6.0 libclang-common-6.0-dev libclang1-6.0 libllvm6.0
           sudo apt install -y clang lld
       - name: Configure
         run: |
@@ -133,8 +131,6 @@ jobs:
       - uses: actions/checkout@v2
       - name: Install Dependencies
         run: |
-          # remove broken packages installed by default
-          sudo apt remove -y clang-6.0 libclang-common-6.0-dev libclang1-6.0 libllvm6.0
           sudo apt install -y clang clang-tidy lld
       - name: Configure
         run: |

--- a/doc/UserGuide.md
+++ b/doc/UserGuide.md
@@ -198,7 +198,7 @@ User defined templates can be adapted for logging in a similar way:
 
     [catchfile test/integration/LoggingAdaptedStructs.cpp adapt_template]
 
-As of C++20, concepts can be also adapted for logging:
+As of C++20, concepts can be also adapted for logging (namespaced concepts are also supported):
 
     [catchfile test/integration/LoggingAdaptedConcepts.cpp concept]
 

--- a/include/binlog/PrettyPrinter.cpp
+++ b/include/binlog/PrettyPrinter.cpp
@@ -61,7 +61,7 @@ void printTwoDigits(binlog::detail::OstreamBuffer& out, int i)
 void printNineDigits(binlog::detail::OstreamBuffer& out, int i)
 {
   char buf[16];
-  snprintf(buf, sizeof(buf), "%.9d", i);
+  (void)snprintf(buf, sizeof(buf), "%.9d", i);
   out.write(buf, 9);
 }
 

--- a/include/binlog/adapt_struct.hpp
+++ b/include/binlog/adapt_struct.hpp
@@ -149,8 +149,8 @@
  * @see BINLOG_ADAPT_STRUCT
  */
 #define BINLOG_ADAPT_CONCEPT(Concept, ...)                                                                         \
-  MSERIALIZE_EXPAND(MSERIALIZE_MAKE_TEMPLATE_TAG((Concept Concept), (Concept) __VA_OPT__(,) __VA_ARGS__))          \
-  MSERIALIZE_EXPAND(MSERIALIZE_MAKE_TEMPLATE_SERIALIZABLE((Concept Concept), (Concept) __VA_OPT__(,) __VA_ARGS__)) \
+  MSERIALIZE_EXPAND(MSERIALIZE_MAKE_TEMPLATE_TAG_I((Concept C), (C), (Concept) __VA_OPT__(,) __VA_ARGS__))         \
+  MSERIALIZE_EXPAND(MSERIALIZE_MAKE_TEMPLATE_SERIALIZABLE((Concept C), (C) __VA_OPT__(,) __VA_ARGS__)) \
   /**/
 #endif // >= C++20
 

--- a/include/mserialize/StructSerializer.hpp
+++ b/include/mserialize/StructSerializer.hpp
@@ -105,41 +105,43 @@ struct StructSerializer
   }
 
 private:
-  template <typename Field, typename OutputStream>
-  static void serialize_member(const T& t, Field T::*field, OutputStream& ostream)
+  // U might be the base of T, if T got adopted through a concept
+
+  template <typename U, typename Field, typename OutputStream>
+  static void serialize_member(const T& t, Field U::*field, OutputStream& ostream)
   {
     mserialize::serialize(t.*field, ostream);
   }
 
-  template <typename Field, typename OutputStream>
-  static void serialize_member(const T& t, Field (T::*getter)() const, OutputStream& ostream)
+  template <typename U, typename Field, typename OutputStream>
+  static void serialize_member(const T& t, Field (U::*getter)() const, OutputStream& ostream)
   {
     mserialize::serialize((t.*getter)(), ostream);
   }
 
 #if __cplusplus >= 201703L
-  template <typename Field, typename OutputStream>
-  static void serialize_member(const T& t, Field (T::*getter)() const noexcept, OutputStream& ostream)
+  template <typename U, typename Field, typename OutputStream>
+  static void serialize_member(const T& t, Field (U::*getter)() const noexcept, OutputStream& ostream)
   {
     mserialize::serialize((t.*getter)(), ostream);
   }
 #endif
 
-  template <typename Field>
-  static std::size_t serialized_size_member(const T& t, Field T::*field)
+  template <typename U, typename Field>
+  static std::size_t serialized_size_member(const T& t, Field U::*field)
   {
     return mserialize::serialized_size(t.*field);
   }
 
-  template <typename Field>
-  static std::size_t serialized_size_member(const T& t, Field (T::*getter)() const)
+  template <typename U, typename Field>
+  static std::size_t serialized_size_member(const T& t, Field (U::*getter)() const)
   {
     return mserialize::serialized_size((t.*getter)());
   }
 
 #if __cplusplus >= 201703L
-  template <typename Field>
-  static std::size_t serialized_size_member(const T& t, Field (T::*getter)() const noexcept)
+  template <typename U, typename Field>
+  static std::size_t serialized_size_member(const T& t, Field (U::*getter)() const noexcept)
   {
     return mserialize::serialized_size((t.*getter)());
   }

--- a/include/mserialize/make_template_tag.hpp
+++ b/include/mserialize/make_template_tag.hpp
@@ -60,16 +60,20 @@
  *     template <typename, typename>
  *     friend struct mserialize::CustomTag;
  */
-//
+#define MSERIALIZE_MAKE_TEMPLATE_TAG(TemplateArgs, ...)                                    \
+  MSERIALIZE_MAKE_TEMPLATE_TAG_I(TemplateArgs, MSERIALIZE_FIRST(__VA_ARGS__), __VA_ARGS__) \
+  /**/
+
 // The macros below deserve a bit of an explanation.
 //
 // Given the following example call:
 //
-// MSERIALIZE_MAKE_TEMPLATE_TAG((typename A, typename B, typename C), (Tuple<A,B,C>), a, b, c)
+// MSERIALIZE_MAKE_TEMPLATE_TAG_I((typename A, typename B, typename C), (Tuple<A,B,C>), (Tuple<A,B,C>), a, b, c)
 //
 // Then the below macros expand to:
 //
 // TemplateArgs = (typename A, typename B, typename C)
+// Specname = (Tuple<A,B,C>)
 // __VA_ARGS__ = (Tuple<A,B,C>), a, b, c
 //               ^-- this is here to avoid empty pack if there are no members,
 //                   as until C++20 the pack cannot be empty
@@ -78,18 +82,20 @@
 // MSERIALIZE_UNTUPLE((Tuple<A,B,C>)) = Tuple<A,B,C>
 // MSERIALIZE_STRINGIZE_LIST(Tuple<A,B,C>) ~= "Tuple<A,B,C>"
 //
+// Specname is useful for concepts, where it is distinct from the typename.
+//
 // tag_guard: to make has_tag report correctly if T<A> has a tag or not,
 // (that depends on the template parameter), we first check if every member
 // has a tag (see conjunction), then either create a constructible member
 // (true_type, if every member has a tag), or a not-constructible member,
 // (BuiltinTag), making this spec also non-constructible - has_tag
 // checks constructibility.
-#define MSERIALIZE_MAKE_TEMPLATE_TAG(TemplateArgs, ...)           \
+#define MSERIALIZE_MAKE_TEMPLATE_TAG_I(TemplateArgs, Specname, ...)           \
   namespace mserialize {                                          \
   template <MSERIALIZE_UNTUPLE(TemplateArgs)>                     \
-  struct CustomTag<MSERIALIZE_UNTUPLE(MSERIALIZE_FIRST(__VA_ARGS__)), void> \
+  struct CustomTag<MSERIALIZE_UNTUPLE(Specname), void>            \
   {                                                               \
-    using T = MSERIALIZE_UNTUPLE(MSERIALIZE_FIRST(__VA_ARGS__));  \
+    using T = MSERIALIZE_UNTUPLE(Specname);                       \
     std::conditional_t<                                           \
       mserialize::detail::conjunction<                            \
         MSERIALIZE_FOREACH(MSERIALIZE_HAS_TAG, _, __VA_ARGS__)    \

--- a/test/integration/LoggingAdaptedConcepts.cpp
+++ b/test/integration/LoggingAdaptedConcepts.cpp
@@ -81,6 +81,8 @@ struct NsFoo {
   std::string ns_str() const { return "bar"; }
 };
 
+struct DerivedNsFoo : NsFoo {};
+
 int main()
 {
   BINLOG_INFO("{}", Nothing{});
@@ -98,6 +100,9 @@ int main()
   // Outputs: Wrap{ a: 456 }
 
   BINLOG_INFO("{}", NsFoo{});
+  // Outputs: ns::Stringable{ ns_str: bar }
+
+  BINLOG_INFO("{}", DerivedNsFoo{});
   // Outputs: ns::Stringable{ ns_str: bar }
 
   binlog::consume(std::cout);

--- a/test/integration/LoggingAdaptedConcepts.cpp
+++ b/test/integration/LoggingAdaptedConcepts.cpp
@@ -65,6 +65,22 @@ BINLOG_ADAPT_TEMPLATE((typename A), (Wrap<A>), a)
 
 static_assert(Stringable<Wrap<int>>);
 
+namespace ns {
+
+template<typename T>
+concept Stringable = requires(T a)
+{
+  { a.ns_str() } -> std::convertible_to<std::string>;
+};
+
+}
+
+BINLOG_ADAPT_CONCEPT(ns::Stringable, ns_str)
+
+struct NsFoo {
+  std::string ns_str() const { return "bar"; }
+};
+
 int main()
 {
   BINLOG_INFO("{}", Nothing{});
@@ -80,6 +96,9 @@ int main()
 
   BINLOG_INFO("{}", Wrap<int>{456});
   // Outputs: Wrap{ a: 456 }
+
+  BINLOG_INFO("{}", NsFoo{});
+  // Outputs: ns::Stringable{ ns_str: bar }
 
   binlog::consume(std::cout);
   return 0;

--- a/test/integration/LoggingBoostTypes.cpp
+++ b/test/integration/LoggingBoostTypes.cpp
@@ -24,6 +24,9 @@ namespace mserialize { namespace detail {
 }}
 //]
 
+// The boost container constructors throw, and clang tidy does not like this
+// NOLINTBEGIN(bugprone-exception-escape)
+
 int main()
 {
   // Boost containers - no specific adoption is required
@@ -74,3 +77,5 @@ int main()
   binlog::consume(std::cout);
   return 0;
 }
+
+// NOLINTEND(bugprone-exception-escape)

--- a/test/perf/LargeLogfile.cpp
+++ b/test/perf/LargeLogfile.cpp
@@ -11,7 +11,7 @@ int main()
   BINLOG_INFO("Hello Large Logfile");
 
   std::size_t size = 0;
-  std::size_t targetSize = 300 * 1024 * 1024;
+  std::size_t targetSize = 300UL * 1024UL * 1024UL;
 
   const std::vector<int> numbers{1,2,3,4,5,6,7,8};
 

--- a/test/unit/binlog/test_utils.cpp
+++ b/test/unit/binlog/test_utils.cpp
@@ -56,7 +56,7 @@ std::string timePointToString(std::chrono::system_clock::time_point tp)
   char buffer[128] = {0};
   const std::time_t tt = std::chrono::system_clock::to_time_t(tp);
   const std::tm* tm = std::localtime(&tt); // NOLINT(concurrency-mt-unsafe)
-  strftime(buffer, 128, "%Y.%m.%d %H:%M:%S", tm);
+  (void)strftime(buffer, 128, "%Y.%m.%d %H:%M:%S", tm);
   return buffer;
 }
 

--- a/test/unit/mserialize/documentation.cpp
+++ b/test/unit/mserialize/documentation.cpp
@@ -1,6 +1,10 @@
 // The purpose of this suite is to test snippets which are included in the documentation automatically.
 // This suite is not to test the features themselves, which are tested in other suites.
 
+// Different snippets that are extracted automatically into the doc
+// include the same headers, ignore them:
+// NOLINTBEGIN(readability-duplicate-include)
+
 #include <doctest/doctest.h>
 
 #include <sstream>
@@ -188,7 +192,7 @@ TEST_CASE("roundtrip")
     //]
   }
 
-  std::remove(path.data());
+  (void)std::remove(path.data());
   CHECK(true);
 }
 
@@ -243,6 +247,8 @@ TEST_CASE("visit")
     //]
   }
 
-  std::remove(path.data());
+  (void)std::remove(path.data());
   CHECK(true);
 }
+
+// NOLINTEND(readability-duplicate-include)

